internal/supervisor/output.go

```diff
@@ -54,10 +55,17 @@ func (m *multiOutput) PipeOutput(proc *process) {
 	pipe := m.openPipe(proc)
 
 	go func(proc *process, pipe *ptyPipe) {
-		scanner := bufio.NewScanner(pipe.pty)
-
-		for scanner.Scan() {
-			m.WriteLine(proc, scanner.Bytes())
+		reader := bufio.NewReader(pipe.pty)
+		for {
+			line, err := reader.ReadBytes('\n')
+			if err != nil {
+				// EOF is expected, so don't log it
+				if err != io.EOF {
+					log.Printf("reader error: %v", err)
+				}
+				break
+			}
+			m.WriteLine(proc, line)
```

Comment:
According to the docs, `line` may be non-empty even in case of an error (e.g. if the last output line is not terminated by `\n`), so may be worth piping it through regardless:

```suggestion
			line, err := reader.ReadBytes('\n')
			if len(line) > 0 {
				m.WriteLine(proc, line)	
			}	
			if err != nil {
				// EOF is expected, so don't log it
				if err != io.EOF {
					log.Printf("reader error: %v", err)
				}
				break
			}

```

Reply:
And one more thing, `ReadBytes` keeps the terminator in the returned `line`, so `WriteLine` may need to account for it:
```
 	buf.WriteString("\033[0m | ")
 
 	buf.Write(p)
-	buf.WriteByte('\n')
+	if p[len(p)-1] != '\n' {
+		buf.WriteByte('\n')
+	}
 
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
```